File: src/ClientFunction.ts
import {
	getRemoteOrThrow,
	IS_CLIENT,
	functionExists,
	waitForFunction,
	MAX_CLIENT_WAITFORCHILD_TIMEOUT,
} from "./internal";

/**
 * A function on the client
 * @rbxts client
 */
export default class NetClientFunction<SR extends unknown> {
	/** @internal */
	private lastPing = -1;
	/** @internal */
	private cached: unknown = [];
	/** @internal */
	private instance: RemoteFunction;

	constructor(name: string) {
		this.instance = getRemoteOrThrow("RemoteFunction", name);
		assert(IS_CLIENT, "Cannot create a Net.ClientFunction on the Server!");
		assert(functionExists(name), `The specified function '${name}' does not exist!`);
	}

	public static async WaitFor<R extends any>(name: string): Promise<NetClientFunction<R>> {
		const fun: RemoteFunction | undefined = waitForFunction(name, MAX_CLIENT_WAITFORCHILD_TIMEOUT);
		if (!fun) {
			error("Failed to retrieve client Function!");
		}

		return new NetClientFunction<R>(name);
	}

	/**
	 * The remoteFunction instance
	 */
	public getInstance() {
		return this.instance;
	}

	/**
	 * The client cache in seconds
	 */
	public getCache() {
		const cache = this.instance.FindFirstChild("Cache") as NumberValue;
		if (cache) {
			return cache.Value;
		} else {
			return 0;
		}
	}

	/**
	 * Call the server with the specified arguments
	 * @param args The arguments to call the server with
	 * @returns the result of the call to the server
	 */
	public CallServer<T extends Array<any>>(...args: T): SR {
		if (this.lastPing < os.time() + this.getCache()) {
			const result = this.instance.InvokeServer(...args);
			this.cached = result;

			this.lastPing = os.time();

			return result as any;
		} else {
			return this.cached as SR;
		}
	}

	/**
	 * Call the server with the specified arguments asynchronously
	 * @param args The args to call the server with
	 * @async Will return a promise
	 */
	public async CallServerAsync<T extends Array<any>>(...args: T): Promise<SR> {
		return this.CallServer(...args);
	}
}

File: ClientFunction.lua

local TS = _G[script]
local _0 = TS.import(script, script.Parent, "internal")
local getRemoteOrThrow = _0.getRemoteOrThrow
local IS_CLIENT = _0.IS_CLIENT
local functionExists = _0.functionExists
local waitForFunction = _0.waitForFunction
local MAX_CLIENT_WAITFORCHILD_TIMEOUT = _0.MAX_CLIENT_WAITFORCHILD_TIMEOUT
--[[
	*
	* A function on the client
	* @rbxts client
]]
local NetClientFunction
do
	NetClientFunction = setmetatable({}, {
		__tostring = function()
			return "NetClientFunction"
		end,
	})
	NetClientFunction.__index = NetClientFunction
	function NetClientFunction.new(...)
		local self = setmetatable({}, NetClientFunction)
		self:constructor(...)
		return self
	end
	function NetClientFunction:constructor(name)
		self.lastPing = -1
		self.cached = {}
		self.instance = getRemoteOrThrow("RemoteFunction", name)
		assert(IS_CLIENT, "Cannot create a Net.ClientFunction on the Server!")
		assert(functionExists(name), "The specified function '" .. name .. "' does not exist!")
	end
	NetClientFunction.WaitFor = TS.async(function(self, name)
		local fun = waitForFunction(name, MAX_CLIENT_WAITFORCHILD_TIMEOUT)
		if not fun then
			error("Failed to retrieve client Function!")
		end
		return NetClientFunction.new(name)
	end)
	function NetClientFunction:Unmanaged(name)
		return getRemoteOrThrow("RemoteFunction", name)
	end
	function NetClientFunction:GetInstance()
		return self.instance
	end
	function NetClientFunction:GetCache()
		local cache = self.instance:FindFirstChild("Cache")
		if cache then
			return cache.Value
		else
			return 0
		end
	end
	function NetClientFunction:CallServer(...)
		local args = { ... }
		if self.lastPing < os.time() + self:GetCache() then
			local result = self.instance:InvokeServer(unpack(args))
			self.cached = result
			self.lastPing = os.time()
			return result
		else
			return self.cached
		end
	end
	NetClientFunction.CallServerAsync = TS.async(function(self, ...)
		local args = { ... }
		return self:CallServer(unpack(args))
	end)
end
return {
	default = NetClientFunction,
}
